internal/controllers/clusterextension_controller.go

```diff
@@ -202,21 +178,133 @@ func (r *ClusterExtensionReconciler) reconcile(ctx context.Context, ext *ocv1alp
+		}
+		if versionRange != "" {
+			return nil, fmt.Errorf("%sno package %q matching version %q found", upgradeErrorPrefix, packageName, versionRange)
+		}
+		if channelName != "" {
+			return nil, fmt.Errorf("%sno package %q found in channel %q", upgradeErrorPrefix, packageName, channelName)
+		}
+		return nil, fmt.Errorf("%sno package %q found", upgradeErrorPrefix, packageName)
+	}
+	sort.SliceStable(resultSet, func(i, j int) bool {
+		return catalogsort.ByVersion(resultSet[i], resultSet[j])
+	})
+	sort.SliceStable(resultSet, func(i, j int) bool {
+		return catalogsort.ByDeprecated(resultSet[i], resultSet[j])
+	})
+
+	return resultSet[0], nil
+}
+
+func (r *ClusterExtensionReconciler) installedBundle(ctx context.Context, allBundles []*catalogmetadata.Bundle, ext *ocv1alpha1.ClusterExtension) (*catalogmetadata.Bundle, error) {
+	bd := &rukpakv1alpha2.BundleDeployment{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: ext.GetName()}, bd)
+	if client.IgnoreNotFound(err) != nil {
+		return nil, err
+	}
+
+	if bd.Spec.Source.Image == nil || bd.Spec.Source.Image.Ref == "" {
+		// Bundle not yet installed
+		return nil, nil
+	}
+
+	bundleImage := bd.Spec.Source.Image.Ref
+	// find corresponding bundle for the installed content
+	resultSet := catalogfilter.Filter(allBundles, catalogfilter.And(
+		catalogfilter.WithPackageName(ext.Spec.PackageName),
+		catalogfilter.WithBundleImage(bundleImage),
+	))
+	if len(resultSet) == 0 {
+		return nil, fmt.Errorf("bundle with image %q for package %q not found in available catalogs but is currently installed via BundleDeployment %q", bundleImage, ext.Spec.PackageName, bd.Name)
+	}
+
+	sort.SliceStable(resultSet, func(i, j int) bool {
+		return catalogsort.ByVersion(resultSet[i], resultSet[j])
+	})
+
+	return resultSet[0], nil
+}
+
+func upgradePredicate(installedBundle *catalogmetadata.Bundle) (catalogfilter.Predicate[catalogmetadata.Bundle], error) {
+	var successors SuccessorsPredicateFunc = LegacySemanticsSuccessorsPredicate
+	if features.OperatorControllerFeatureGate.Enabled(features.ForceSemverUpgradeConstraints) {
+		successors = SemverSuccessorsPredicate
+	}
+
+	installedBundleVersion, err := installedBundle.Version()
+	if err != nil {
+		return nil, err
+	}
+
+	installedVersionConstraint, err := mmsemver.NewConstraint(installedBundleVersion.String())
+	if err != nil {
 		return nil, err
 	}
-	bundleDeploymentList := rukpakv1alpha2.BundleDeploymentList{}
-	if err := r.Client.List(ctx, &bundleDeploymentList); err != nil {
+
+	successorsPredicate, err := successors(installedBundle)
+	if err != nil {
 		return nil, err
 	}
 
-	return GenerateVariables(allBundles, clusterExtensionList.Items, bundleDeploymentList.Items)
+	// We need either successors or current version (no upgrade)
+	return catalogfilter.Or(
+		successorsPredicate,
+		catalogfilter.InMastermindsSemverRange(installedVersionConstraint),
```

Comment:
This needs to be something like this:

```go
catalogfilter.And(
	catalogfilter.WithPackageName(installedBundle.Package),
	catalogfilter.InMastermindsSemverRange(installedVersionConstraint),
)
```

Otherwise we will include bundles which match `installedVersionConstraint`, but not from the same package.